---
title: background
layout: default
---

# The eGFRD algorithm
A reaction-diffusion system is a many-body problem that cannot be solved analytically. The key idea of eGFRD is to decompose the many-body problem into one- and two-body problems that can be solved analytically via Green’s functions, and used to set up the algorithm \[[1,2](#references)\]. The Green’s functions allow eGFRD to make large jumps in time and space when the particles are far apart from each other.
In the original version of the algorithm (GFRD), the many-body problem was solved by determining at each step of the simulation a maximum time step such that each particle could interact with at most one other particle during that time step \[[1,2](#references)\]. This scheme was a synchronous event-driven algorithm, because at each time step all the particles were propagated simultaneously. Moreover, the scheme was not exact, because the decomposition into single particles and pairs of particles involved cut-off distances, introducing a trade-off between speed and error.

In the current version (eGFRD) we have implemented ideas of Opplestrup and coworkers \[[3](#references)\]. In this new scheme, protective domains are put around single particles and pairs of particles, as shown in Fig. 1. For each of the domains, the reaction-diffusion problem is solved analytically using Green’s functions. This yields for each domain an event type and an event time; as described below, the set of possible event types depends on whether the domain is a Single, meaning it contains a single particle, or a Pair, meaning it contains a pair of particles. The event times are put in a chronologically ordered event list, and the events are then executed in chronological order. When an event is executed, the particles of the corresponding domain are propagated, for the propagated particles new domains, with new event types and new event times, are determined, and the new events are put back in the event list. The algorithm is thus exact, event-driven and asynchronous. Its asynchronous nature makes eGFRD similar in spirit to event-driven MD simulations of hard spheres and the Gibson-Bruck scheme, which is an exact, event-driven asynchronous algorithm for simulating the zero-dimensional chemical master equation \[[4](#references)\].

<p align="center"><img src="includes/images/gfrd.svg" alt="Overview of GFRD"/></p>
<div style="margin:auto;width:75%;text-align:center;font-style:italic">
Fig.1: Overview of eGFRD. The principal idea of GFRD is to decompose the many-body reaction-diffusion problem into one- and two-body problems that can be solved analytically using Green’s Functions. To this end, protective domains are put around single particles and pairs of particles, leading to so-called Singles and Pairs, respectively. For each domain, a next event type and a next event time is determined. The event times are put in a chronologically ordered event list, and the events are then executed in chronological order. When an event is executed, the particles of the corresponding domain are propagated, new domains with new events are determined, and the events are put back in the event list.
</div>

## Single
A Single is a domain that contains only a single particle. The next event type is either a unimolecular reaction of the type A ⇾ B + C + … or A ⇾ Ø, or an escape from the domain, as sketched in Fig. 2. To determine which of the two possible events will occur, a tentative event time is determined for each of them; the event with the smallest tentative event time is the one that will occur for that domain; it is the one that is put in the event list.

<p align="center"><img src="includes/images/single.svg" alt="A single"/></p>
<div style="margin:auto;width:75%;text-align:center;font-style:italic">
Fig.2: A Single, a protective domain that contains only a single particle. The possible event types are either a unimolecular reaction or the escape of the single particle from the protective domain.
</div>

## Pair
A Pair is a domain that contains a pair of particles. The reaction-diffusion problem of two particles that can react with each other and diffuse in a spherical domain with absorbing boundary conditions can, to our knowledge, not be solved analytically. We therefore apply a coordinate transformation leading to a diffusion problem for the center-of-mass coordinate R and a reaction-diffusion problem for the inter-particle vector r, as sketched in Fig. 3. The event types that can occur are R leaving its domain, r leaving its domain, a unimolecular reaction or a bimolecular reaction. For each of these events a tentative event time is determined, and the event with the smallest tentative event time is the one that will happen, and put in the event list.

<p align="center"><img src="includes/images/pair.svg" alt="A pair"/></p>
<div style="margin:auto;width:75%;text-align:center;font-style:italic">
Fig.3: A Pair, a protective domain that contains a pair of particles. A coordinate transformation is applied, and one protective domain is defined for the center-of-mass coordinate <b>R</b> and one for the inter-particle vector <b>r</b>. The sizes of these domains are determined such that when both <b>R</b> and <b>r</b> would reach their maximum values, |<b>R</b>|<sup>max</sup> and |<b>r</b>|<sup>max</sup>, respectively, the two particles would still be within the original protective domain. The dynamics of <b>R</b> is a diffusion problem of a random walker in a spherical domain with absorbing boundary conditions. The dynamics of <b>r</b> takes into account that the two particles not only diffuse, but can also react with each other; this is the problem of a random walker between two concentric spherical surfaces, with a radiation boundary condition at the inner surface, describing the bimolecular reaction, and an adsorbing boundary condition at the outer surface, describing the escape from the domain. The next event is thus either the escape of <b>R</b> from its domain, the escape of <b>r</b> from its domain, a bimolecular reaction, or a unimolecular reaction.
</div>

# The importance of modeling networks
In many networks of interacting components, the spatial distribution of the components and the stochastic character of their interactions are of critical importance for the macroscopic behavior of the system. Examples are to be found in biological evolution and population dynamics, but arguably the most studied and best characterized examples are biochemical networks. Biochemical networks consist of biomolecules, such as proteins and DNA, which chemically and physically interact with one another. These biochemical networks are the analog computers of life. They allow living cells to detect, amplify and integrate signals, as well as transmit signals from one place to another. Indeed, biochemical networks can perform a large number of computational tasks analogous to electronic circuits. However, their design principles are markedly different. In a biochemical network, the components often move in an erratic fashion, namely by diffusion, and they interact with each other in a stochastic manner – chemical and physical interactions are probabilistic in nature. To understand the design principles that allow biochemical networks to process information reliably, we have to take into account the spatial distribution of the components and the stochastic character of their interactions \[[5](#references)\]. Moreover, we have recently shown that spatio-temporal correlations at molecular length scales can drastically change the macroscopic behavior of systems that are uniform at cellular length scales \[[6](#references)\]. Finally, many biological systems exhibit macroscopic concentration gradients. Well-known examples are morphogen gradients, which provide positional information to differentiating cells during embryonic development. To understand how these gradients are read out precisely \[[7](#references)\], the system has to be described at the particle level in time and space \[[8](#references)\].

## Overview of simulation techniques
In principle, computer simulations are ideally suited for studying reaction-diffusion systems. The conventional approach is to write down the macroscopic reaction-diffusion equations and solve the corresponding differential equations either analytically or numerically. This approach resolves the system in time and space, but ignores the discrete nature of the reactants and the stochastic character of their interactions. Schemes based on the zero-dimensional chemical master equation, such as the Gillespie algorithm \[[9](#references)\] or the Gibson-Bruck scheme \[[10](#references)\], incorporate the latter, but assume that at each instant the particles are uniformly distributed in space.
In the past years, a number of techniques have been developed that take into account both the spatial distribution of the components and the stochastic nature of their interactions. One class of techniques is based upon the reaction-diffusion master equation \[[11, 12](#references)\]; examples are SmartCell \[[13](#references)\], MesoRD \[[14](#references)\] and URMD \[[15](#references)\]. The main idea of these techniques is to divide the reaction volume into a number of subvolumes; particles can react within the subvolumes, but also diffuse from one subvolume to the next. Importantly, it is assumed that within each subvolume the particles are well mixed. These techniques thus rely on the existence of a length and time scale on which the system is spatially uniform. When the concentrations are fairly high, as in spatio-temporal calcium oscillations, such a scale possibly exists, but at lower concentrations this assumption is likely to fail.
Another class of techniques simulates the network in time and space at the particle level. One natural approach is to use Brownian Dynamics; examples are Smoldyn \[[16](#references)\], MCell \[[17](#references)\], Reaction Brownian Dynamics \[[18](#references)\] and GridCell \[[19](#references)\]. In essence, the particles undergo a random walk, and when two reaction partners happen to meet each other, they can react with a probability consistent with the rate constant. However, under biologically relevant conditions, Brownian Dynamics is not very efficient, since much CPU time is wasted on propagating the particles towards each other. Moreover, it is not exact, since a finite time step is used.
Another approach to simulate a network at the particle level in time and space is to use an event-driven algorithm. Green’s Function Reaction Dynamics (GFRD) \[[6, 20, 21](#references)\] and First Passage Kinetic Monte Carlo (FPKMC) \[[22, 23](#references)\] are examples of such an approach. The idea is to exploit the solution of the Smoluchowski equation — the Green’s function — to set up an event-driven algorithm that concatenates the propagation of the particles in space with the reactions between them. These algorithms thus alleviate the need to propagate the particles toward each other to let them react, as in conventional Brownian Dynamics: even when the reactants are far apart from one another, the algorithm can immediately jump to the next reaction event. This event-driven nature makes these schemes highly powerful, especially when the concentrations of the components are low, as in most signal transduction pathways and gene regulation networks.

# References
1. van Zon JS, tenWolde PR (2005) Simulating biochemical networks at the particle level and in time and space: Green’s function reaction dynamics. Phys Rev Lett, 94:128103. ([doi](https://dx.doi.org/10.1103/PhysRevLett.94.128103)))
2. van Zon JS, ten Wolde PR (2005) Green’s-function reaction dynamics: A particle-based approach for simulating biochemical networks in time and space. J Chem Phys, 123:234910. ([doi](https://dx.doi.org/10.1063/1.2137716), [arXiv](https://arxiv.org/abs/q-bio/0404002))
3. Opplestrup T, Bulatov VV, Gilmer GH, Kalos MH, Sadigh B (2006) First-passage Monte Carlo algorithm: diffusion without all the hops. Phys Rev Lett, 97:230602. ([doi](https://dx.doi.org/10.1103/PhysRevLett.97.230602), [arXiv](https://arxiv.org/abs/0905.3576))
4. Gibson MA, Bruck J (2000) Efficient exact stochastic simulation of chemical systems with many species and many channels. J Phys Chem A, 104: 1876 — 1888. ([doi](https://dx.doi.org/10.1021/jp993732q))
5. van Zon JS, Morelli M, Tanase-Nicola S, ten Wolde PR (2006) Diffusion of transcription factors can drastically enhance the noise in gene expression. Biophys J, 91:4350 — 4367. ([doi](https://dx.doi.org/10.1529/biophysj.106.086157), [arXiv](https://arxiv.org/abs/q-bio/0604005))
6. Takahashi K, Tanase-Nicola S, Ten Wolde PR (2010), Spatio-temporal correlations can drastically change the response of a MAPK pathway. Proc Natl. Acad Sci USA, 107:2473 — 2478. ([doi](https://dx.doi.org/10.1073/pnas.0906885107), [arXiv](https://arxiv.org/abs/0907.0514))
7. Gregor T, Tank DW, Wieschaus EF, Bialek W (2007) Probing the limits to positional information. Cell, 130: 153 — 164. ([doi](https://dx.doi.org/10.1016/j.cell.2007.05.025))
8. Erdmann T, Howard M, Ten Wolde PR (2009) Role of spatial averaging in the precision of gene expression patterns. Phys Rev Lett, 103: 258101. ([doi](https://dx.doi.org/10.1103/PhysRevLett.103.258101), [arXiv](https://arxiv.org/abs/0907.4289))
9. Gillespie DT (1977) Exact stochastic simulation of coupled chemical reactions. J Phys Chem, 81:2340 — 2361. ([doi](https://dx.doi.org/10.1021/j100540a008))
10. Gibson MA, Bruck J (2000) Efficient exact stochastic simulation of chemical systems with many species and many channels. J Phys Chem A, 104: 1876 — 1888. ([doi](https://dx.doi.org/10.1021/jp993732q))
11. Baras F, Mansour M (1996) Reaction-diffusion master equation: A comparison with microscopic simulations. Phys Rev E, 54:6139 — 6148. ([doi](https://dx.doi.org/10.1103/PhysRevE.54.6139))
12. Stundzia AB, Lumsden CJ (1996) Stochastic simulation of coupled reaction-diffusion processes. J Comp Phys, 127: 196 — 207. ([doi](https://dx.doi.org/10.1006/jcph.1996.0168))
13. Ander M et al. (2004) SmartCell, a framework to simulate cellular processes that combines stochastic approximation with diffusion and localization: analysis of simple networks. Sys Biol, 1:129 — 138. ([doi](https://dx.doi.org/10.1049/sb:20045017))
14. Hattne J, Fange D, Elf J (2005) Stochastic reaction-diffusion simulation with MesoRD. Bioinformatics, 21: 2923 — 2924. ([doi](https://dx.doi.org/10.1093/bioinformatics/bti431))
15. Engblom S, Ferm L, Hellander A, Lötstedt P (2009) Simulation of stochastic reaction-diffusion processes on unstructured meshes. SIAM J Sci Comput, 31:1774 — 1797. ([doi](https://dx.doi.org/10.1137/080721388), [arXiv](https://arxiv.org/abs/0804.3288))
16. Andrews S, Bray D (2004) Stochastic simulation of chemical reactions with spatial resolution and single molecule detail. Phys Biol, 1:137 — 151. ([doi](https://dx.doi.org/10.1088/1478-3967/1/3/001))
17. Coggan JS, Bartol TM, Esquenazi E, Stiles JR, Lamont S, Martone ME, Berg DK, Ellisman MH, Sejnowski TJ (2005) Evidence for ectopic neurotransmission at a neauronal synapse. Science, 309:446 — 451. ([doi](https://dx.doi.org/10.1126/science.1108239))
18. Morelli MJ, Ten Wolde PR (2008) Reaction Brownian Dynamics and the effect of spatial fluctuations on the gain of a push-pull network. J Chem Phys, 129: 054112. ([doi](https://dx.doi.org/10.1063/1.2958287), [arXiv](https://arxiv.org/abs/0804.4125))
19. Boulianne L, Al Assaad S, Dumontier M, Gross WJ (2008) GridCell: a stochastic particle-based biological system simulator. BMC Sys Biol, 2:66. ([doi](https://dx.doi.org/10.1186/1752-0509-2-66))
20. van Zon JS, tenWolde PR (2005) Simulating biochemical networks at the particle level and in time and space: Green’s function reaction dynamics. Phys Rev Lett, 94:128103. ([doi](https://dx.doi.org/10.1103/PhysRevLett.94.128103))
21. van Zon JS, ten Wolde PR (2005) Green’s-function reaction dynamics: A particle-based approach for simulating biochemical networks in time and space. J Chem Phys, 123:234910. ([doi](https://dx.doi.org/10.1063/1.2137716), [arXiv](https://arxiv.org/abs/q-bio/0404002))
22. Opplestrup T, Bulatov VV, Gilmer GH, Kalos MH, Sadigh B (2006) First-passage Monte Carlo algorithm: diffusion without all the hops. Phys Rev Lett, 97:230602. ([doi](https://dx.doi.org/10.1103/PhysRevLett.97.230602), [arXiv](https://arxiv.org/abs/0905.3576))
23. Opplestrup T, Bulatov VV, Donev A, Kalos MH, Gilmer GH, Sadigh B (2009) First passage kinetic Monte Carlo. Phys Rev E, 80: 066701. ([doi](https://dx.doi.org/10.1103/PhysRevE.80.066701), [arXiv](https://arxiv.org/abs/0905.3575))