---
title: overview
layout: default
---

## The importance of modeling networks in time and space at the particle level
In many networks of interacting components, the spatial distribution of the components and the stochastic character of their interactions are of critical importance for the macroscopic behavior of the system. Examples are to be found in biological evolution and population dynamics, but arguably the most studied and best characterized examples are biochemical networks. Biochemical networks consist of biomolecules, such as proteins and DNA, which chemically and physically interact with one another. These biochemical networks are the analog computers of life. They allow living cells to detect, amplify and integrate signals, as well as transmit signals from one place to another. Indeed, biochemical networks can perform a large number of computational tasks analogous to electronic circuits. However, their design principles are markedly different. In a biochemical network, the components often move in an erratic fashion, namely by diffusion, and they interact with each other in a stochastic manner – chemical and physical interactions are probabilistic in nature. To understand the design principles that allow biochemical networks to process information reliably, we have to take into account the spatial distribution of the components and the stochastic character of their interactions \[[1](#references)\]. Moreover, we have recently shown that spatio-temporal correlations at molecular length scales can drastically change the macroscopic behavior of systems that are uniform at cellular length scales \[[2](#references)\]. Finally, many biological systems exhibit macroscopic concentration gradients. Well-known examples are morphogen gradients, which provide positional information to differentiating cells during embryonic development. To understand how these gradients are read out precisely \[[3](#references)\], the system has to be described at the particle level in time and space \[[4](#references)\].

## Overview of simulation techniques
In principle, computer simulations are ideally suited for studying reaction-diffusion systems. The conventional approach is to write down the macroscopic reaction-diffusion equations and solve the corresponding differential equations either analytically or numerically. This approach resolves the system in time and space, but ignores the discrete nature of the reactants and the stochastic character of their interactions. Schemes based on the zero-dimensional chemical master equation, such as the Gillespie algorithm \[[5](#references)\] or the Gibson-Bruck scheme \[[6](#references)\], incorporate the latter, but assume that at each instant the particles are uniformly distributed in space.

In the past years, a number of techniques have been developed that take into account both the spatial distribution of the components and the stochastic nature of their interactions. One class of techniques is based upon the reaction-diffusion master equation \[[7,8](#references)\]; examples are SmartCell \[[9](#references)\], MesoRD \[[10](#references)\] and URMD \[[11](#references)\]. The main idea of these techniques is to divide the reaction volume into a number of subvolumes; particles can react within the subvolumes, but also diffuse from one subvolume to the next. Importantly, it is assumed that within each subvolume the particles are well mixed. These techniques thus rely on the existence of a length and time scale on which the system is spatially uniform. When the concentrations are fairly high, as in spatio-temporal calcium oscillations, such a scale possibly exists, but at lower concentrations this assumption is likely to fail.

Another class of techniques simulates the network in time and space at the particle level. One natural approach is to use Brownian Dynamics; examples are Smoldyn \[[12](#references)\], Reaction Brownian Dynamics \[[13](#references)\], and GridCell \[[14](#references)\]. In essence, the particles undergo a random walk, and when two reaction partners happen to meet each other, they can react with a probability consistent with the rate constant. However, under biologically relevant conditions, Brownian Dynamics is not very efficient, since much CPU time is wasted on propagating the particles towards each other. Moreover, it is not exact, since a finite time step is used. 

Another approach to simulate a network at the particle level in time and space is to use an event-driven algorithm. Green’s Function Reaction Dynamics (GFRD) \[[2, 14, 15](#references)\] and First Passage Kinetic Monte Carlo (FPKMC) \[[17, 18](#references)\]. are examples of such an approach. The idea is to exploit the solution of the Smoluchowski equation – the Green’s function – to set up an event-driven algorithm that concatenates the propagation of the particles in space with the reactions between them. These algorithms thus alleviate the need to propagate the particles toward each other to let them react, as in conventional Brownian Dynamics: even when the reactants are far apart from one another, the algorithm can immediately jump to the next reaction event. This event-driven nature makes these schemes highly powerful, especially when the concentrations of the components are low, as in most signal transduction pathways and gene regulation networks.

## References
1. van Zon JS, Morelli M, Tanase-Nicola S, ten Wolde PR (2006) Diffusion of transcription factors can drastically enhance the noise in gene expression. Biophys J, 91:4350 — 4367. ([doi](https://dx.doi.org/10.1529/biophysj.106.086157), [arXiv](https://arxiv.org/abs/q-bio/0604005))
2. Takahashi K, Tanase-Nicola S, Ten Wolde PR (2010), Spatio-temporal correlations can drastically change the response of a MAPK pathway. Proc Natl. Acad Sci USA, 107:2473 — 2478. ([doi](https://dx.doi.org/10.1073/pnas.0906885107), [arXiv](https://arxiv.org/abs/0907.0514))
3. Gregor T, Tank DW, Wieschaus EF, Bialek W (2007) Probing the limits to positional information. Cell, 130: 153 — 164. ([doi](https://dx.doi.org/10.1016/j.cell.2007.05.025))
4. Erdmann T, Howard M, Ten Wolde PR (2009) Role of spatial averaging in the precision of gene expression patterns. Phys Rev Lett, 103: 258101. ([doi](https://dx.doi.org/10.1103/PhysRevLett.103.258101), [arXiv](https://arxiv.org/abs/0907.4289))
5. Gillespie DT (1977) Exact stochastic simulation of coupled chemical reactions. J Phys Chem, 81:2340 — 2361. ([doi](https://dx.doi.org/10.1021/j100540a008))
6. Gibson MA, Bruck J (2000) Efficient exact stochastic simulation of chemical systems with many species and many channels. J Phys Chem A, 104: 1876 — 1888. ([doi](https://dx.doi.org/10.1021/jp993732q))
7. Baras F, Mansour M (1996) Reaction-diffusion master equation: A comparison with microscopic simulations. Phys Rev E, 54:6139 — 6148. ([doi](https://dx.doi.org/10.1103/PhysRevE.54.6139))
8. Stundzia AB, Lumsden CJ (1996) Stochastic simulation of coupled reaction-diffusion processes. J Comp Phys, 127: 196 — 207. ([doi](https://dx.doi.org/10.1006/jcph.1996.0168))
9. Ander M et al. (2004) SmartCell, a framework to simulate cellular processes that combines stochastic approximation with diffusion and localization: analysis of simple networks. Sys Biol, 1:129 — 138. ([doi](https://dx.doi.org/10.1049/sb:20045017))
10. Hattne J, Fange D, Elf J (2005) Stochastic reaction-diffusion simulation with MesoRD. Bioinformatics, 21: 2923 — 2924. ([doi](https://dx.doi.org/10.1093/bioinformatics/bti431))
11. Engblom S, Ferm L, Hellander A, Lötstedt P (2009) Simulation of stochastic reaction-diffusion processes on unstructured meshes. SIAM J Sci Comput, 31:1774 — 1797. ([doi](https://dx.doi.org/10.1137/080721388), [arXiv](https://arxiv.org/abs/0804.3288))
12. Andrews S, Bray D (2004) Stochastic simulation of chemical reactions with spatial resolution and single molecule detail. Phys Biol, 1:137 — 151. ([doi](https://dx.doi.org/10.1088/1478-3967/1/3/001))
13. Morelli MJ, Ten Wolde PR (2008) Reaction Brownian Dynamics and the effect of spatial fluctuations on the gain of a push-pull network. J Chem Phys, 129: 054112. ([doi](https://dx.doi.org/10.1063/1.2958287), [arXiv](https://arxiv.org/abs/0804.4125))
14. Boulianne L, Al Assaad S, Dumontier M, Gross WJ (2008) GridCell: a stochastic particle-based biological system simulator. BMC Sys Biol, 2:66. ([doi](https://dx.doi.org/10.1186/1752-0509-2-66))
15. van Zon JS, tenWolde PR (2005) Simulating biochemical networks at the particle level and in time and space: Green’s function reaction dynamics. Phys Rev Lett, 94:128103. ([doi](https://dx.doi.org/10.1103/PhysRevLett.94.128103))
16. van Zon JS, ten Wolde PR (2005) Green’s-function reaction dynamics: A particle-based approach for simulating biochemical networks in time and space. J Chem Phys, 123:234910. ([doi](https://dx.doi.org/10.1063/1.2137716), [arXiv](https://arxiv.org/abs/q-bio/0404002))
17. Opplestrup T, Bulatov VV, Gilmer GH, Kalos MH, Sadigh B (2006) First-passage Monte Carlo algorithm: diffusion without all the hops. Phys Rev Lett, 97:230602. ([doi](https://dx.doi.org/10.1103/PhysRevLett.97.230602), [arXiv](https://arxiv.org/abs/0905.3576))
18. Opplestrup T, Bulatov VV, Donev A, Kalos MH, Gilmer GH, Sadigh B (2009) First passage kinetic Monte Carlo. Phys Rev E, 80: 066701. ([doi](https://dx.doi.org/10.1103/PhysRevE.80.066701), [arXiv](https://arxiv.org/abs/0905.3575))